### ISSUE TITLE
GM_getValue() returns a JSON object, although an array was stored

### ISSUE THREAD

**Issue by double-beep** [Tampermonkey/tampermonkey#1376]
Consider the following script:

```
// ==UserScript==
// @name         Test GM_set/getValue()
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @author       You
// @match        https://example.com/
// @grant        GM_getValue
// @grant        GM_setValue
// ==/UserScript==

(function() {
    'use strict';

    GM_setValue('test',
        [
            {
                name: 'foo',
                value: 'bar'
            },
            {
                name: 'foo1',
                value: 'bar1'
            }
        ]
    );
    console.log(GM_getValue('test'), 'is array:', Array.isArray(GM_getValue('test')));
})();
```

Since I store an array, I expect an array to be logged. However, I get this:

```
{
    "0": {
        "name": "foo",
        "value": "bar"
    },
    "1": {
        "name": "foo1",
        "value": "bar1"
    }
}
```

## Specifications

- TM BETA v4.14.6147

**Comment by jb222** [comment]
Hoping for a quick fix because this is why my userscript no longer works...

**Comment by derjanb** [comment]
Should be fixed at 4.14.6148 (in review|[crx](https://tampermonkey.net/crx/tampermonkey_beta_4_14_6148.crx))

**Comment by filbo** [comment]
My Opera just updated to .6147 and this started happening.  Please assure me that no non-beta TM release has shipped with this bug?

**Comment by derjanb** [comment]
@filbo This means that you've installed Tampermonkey from the Chrome Web Store. Latest stable version is 4.13 and the next will be 4.14.

**Comment by filbo** [comment]
I think I actually have it installed from the Opera 'store', but the beta not release build gibberishid.  Can't check right now, replying from a Dr. waiting room on my phone...  Anyway, I am reassured that only beta channel surfers will experience this, so the system is working as intended...

**Comment by filbo** [comment]
(the release gibberishid in Opera's 'store' is unusable because they 'detect it as malware' for spurious reasons)

**Comment by filbo** [comment]
OK, the unusable (therefore disabled) one is dhdgffkkebhmkfjojejmpbldmpobfkfo, from Opera 'store' main entry -- which I see no longer exists.  The working one is the beta from Chrome 'store', gcalenpjmijncebpfijmoaglllgpjagf.  Anyway, I have a more important issue which I'll put in as a separate comment.

**Comment by filbo** [comment]
I had a failure because an object acquired from GM_getValue() had no .hasOwnProperty(): 'myObject.hasOwnProperty is not a function'.  Taking a closer look -- GM_getValue() isn't just returning plain objects when they should be arrays.  Where they should be objects, it's returning objects with NO PROPERTIES!  `myObject.__proto__` doesn't exist; `myObject.hasOwnProperty` doesn't exist; no properties at all exist, except the ones I had previously saved into GM storage.  It's like the object was created via Object.create(null).

I'm currently postprocessing returns from GM_getValue() with: if it's supposed to be an array, `myArray = Object.values(myArray)`; if it's supposed to be an object, `myObject = JSON.parse(JSON.stringify(myObject))`.  Which is all horrible of course :)

Reporting this just in case the cause is different, or the currently intended fix for arrays would miss this.

**Comment by the1812** [comment]
Same issue with `GM_xmlhttpRequest` in `4.14.6148`.
```ts
GM_xmlhttpRequest({
  url: 'https://jsonplaceholder.typicode.com/todos',
  responseType: 'json',
  onload: resp => {
    console.log(Array.isArray(resp))
    // false in Tampermonkey 4.14.6147+
  },
})
```

**Comment by double-beep** [comment]
@the1812 eh, that's not a bug. This is an object because [it should be](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest). You can use `resp.response` to get the parsed array.

**Comment by the1812** [comment]
@double-beep Sorry, that was a typo, got it fixed now. `resp.response` is still an object.
![image](https://user-images.githubusercontent.com/26504152/144771695-4a80afd1-de9f-48fb-a631-597aa9d3f2bb.png)

**Comment by derjanb** [comment]
> `resp.response` is still an object.

Will be fixed at the next BETA release. Thanks for reporting this.

**Comment by filbo** [comment]
@derjanb can you confirm you saw my (https://github.com/Tampermonkey/tampermonkey/issues/1376#issuecomment-983210388) above?

GM_getValue() isn't just converting arrays to objects.  Values which started out as objects are also coming back strangely transformed.

It's returning objects with _no properties at all_, other than what was in the JSON.  No `.__proto__`, therefore no `.hasOwnProperty()` etc.

**Comment by derjanb** [comment]
@filbo Yes, and this issue should be fixed at 4.14.6148. Isn't it?

**Comment by filbo** [comment]
@derjanb yes, thanks!  My browser was still on 4.14.6147; when I force updated, everything went back to sanity.
